core/src/test/scala/org/apache/spark/resource/ResourceUtilsSuite.scala

```diff
@@ -0,0 +1,230 @@
+      val gpuDiscovery = createTempScriptWithExpectedOutput(dir, "gpuDiscoveryScript",
+        """{"name": "gpu", "addresses": ["0", "1"]}""")
+      conf.set(EXECUTOR_GPU_ID.amountConf, "2")
+      conf.set(EXECUTOR_GPU_ID.discoveryScriptConf, gpuDiscovery)
+
+      val fpgaDiscovery = createTempScriptWithExpectedOutput(dir, "fpgDiscoverScript",
+        """{"name": "fpga", "addresses": ["f1", "f2", "f3"]}""")
+      conf.set(EXECUTOR_FPGA_ID.amountConf, "2")
+      conf.set(EXECUTOR_FPGA_ID.discoveryScriptConf, fpgaDiscovery)
+
+      val resources = getOrDiscoverAllResources(conf, SPARK_EXECUTOR_PREFIX, None)
+      assert(resources.size === 2)
+      val gpuValue = resources.get(GPU)
+      assert(gpuValue.nonEmpty, "Should have a gpu entry")
+      assert(gpuValue.get.name == "gpu", "name should be gpu")
+      assert(gpuValue.get.addresses.size == 2, "Should have 2 indexes")
+      assert(gpuValue.get.addresses.deep == Array("0", "1").deep, "should have 0,1 entries")
+
+      val fpgaValue = resources.get(FPGA)
+      assert(fpgaValue.nonEmpty, "Should have a gpu entry")
+      assert(fpgaValue.get.name == "fpga", "name should be fpga")
+      assert(fpgaValue.get.addresses.size == 3, "Should have 3 indexes")
+      assert(fpgaValue.get.addresses.deep == Array("f1", "f2", "f3").deep,
+        "should have f1,f2,f3 entries")
+    }
+  }
+
+  test("list resource ids") {
+    val conf = new SparkConf
+    conf.set(DRIVER_GPU_ID.amountConf, "2")
+    var resources = listResourceIds(conf, SPARK_DRIVER_PREFIX)
+    assert(resources.size === 1, "should only have GPU for resource")
+    assert(resources(0).resourceName == GPU, "name should be gpu")
+
+    conf.set(DRIVER_FPGA_ID.amountConf, "2")
+    val resourcesMap = listResourceIds(conf, SPARK_DRIVER_PREFIX)
+      .map{ rId => (rId.resourceName, 1)}.toMap
+    assert(resourcesMap.size === 2, "should only have GPU for resource")
+    assert(resourcesMap.get(GPU).nonEmpty, "should have GPU")
+    assert(resourcesMap.get(FPGA).nonEmpty, "should have FPGA")
+  }
+
+  test("parse resource request") {
+    val conf = new SparkConf
+    conf.set(DRIVER_GPU_ID.amountConf, "2")
+    var request = parseResourceRequest(conf, DRIVER_GPU_ID)
+    assert(request.id.resourceName === GPU, "should only have GPU for resource")
+    assert(request.amount === 2, "GPU count should be 2")
+    assert(request.discoveryScript === None, "discovery script should be empty")
+    assert(request.vendor === None, "vendor should be empty")
+
+    val vendor = "nvidia.com"
+    val discoveryScript = "discoveryScriptGPU"
+    conf.set(DRIVER_GPU_ID.discoveryScriptConf, discoveryScript)
+    conf.set(DRIVER_GPU_ID.vendorConf, vendor)
+    request = parseResourceRequest(conf, DRIVER_GPU_ID)
+    assert(request.id.resourceName === GPU, "should only have GPU for resource")
+    assert(request.amount === 2, "GPU count should be 2")
+    assert(request.discoveryScript.get === discoveryScript, "discovery script should be empty")
```

Comment:
"discovery script should be discoveryScriptGPU"

Reply:
changed to "should get discovery script". putting `discoveryScriptGPU` in results duplicate info.